src/Microsoft.Data.Analysis/DataFrame.Join.cs

```diff
@@ -155,207 +164,220 @@ public DataFrame Join(DataFrame other, string leftSuffix = "_left", string right
+                
+                // shrink join result on current column by previouse join columns (if any)
+                // (we have to remove occurrences that doesn't exist in previouse columns, because JOIN happens only if ALL left and right columns in JOIN are matched)
+                if (occurrences != null)
                 {
-                    var otherColumnValue = otherColumn[i];
-                    if (otherColumnValue != null)
-                    {
-                        if (multimap.TryGetValue((TKey)otherColumnValue, out ICollection<long> rowNumbers))
-                        {
-                            foreach (long row in rowNumbers)
-                            {
-                                leftRowIndices.Append(leftColumnIsSmaller ? row : i);
-                                rightRowIndices.Append(leftColumnIsSmaller ? i : row);
-                            }
-                        }
-                    }
-                    else
+                    var shrinkedOccurences = new Dictionary<long, ICollection<long>>();
+
+                    foreach (var kvp in newOccurrences)
                     {
-                        foreach (long nullIndex in smallerDataFrameColumnNullIndices)
+                        var newValue = kvp.Value.Where(i => occurrences[kvp.Key].Contains(i)).ToArray();
+                        if (newValue.Any())
                         {
-                            leftRowIndices.Append(leftColumnIsSmaller ? nullIndex : i);
-                            rightRowIndices.Append(leftColumnIsSmaller ? i : nullIndex);
+                            shrinkedOccurences.Add(kvp.Key, newValue);
                         }
                     }
+                    newOccurrences = shrinkedOccurences;
                 }
-            }
-            else if (joinAlgorithm == JoinAlgorithm.FullOuter)
-            {
-                DataFrameColumn otherColumn = other.Columns[rightJoinColumn];
-                Dictionary<TKey, ICollection<long>> multimap = otherColumn.GroupColumnValues<TKey>(out HashSet<long> otherColumnNullIndices);
-                Dictionary<TKey, long> intersection = new Dictionary<TKey, long>(EqualityComparer<TKey>.Default);
 
-                // Go over the records in this dataframe and match with the dictionary 
-                DataFrameColumn thisColumn = Columns[leftJoinColumn];
-                Int64DataFrameColumn thisColumnNullIndices = new Int64DataFrameColumn("ThisColumnNullIndices");
+                occurrences = newOccurrences;
+            }
+            
+            retainedRowIndices = new Int64DataFrameColumn("RetainedIndices");
+            supplementaryRowIndices = new Int64DataFrameColumn("SupplementaryIndices");
 
-                for (long i = 0; i < thisColumn.Length; i++)
+            //Perform Merging 
+            var retainJoinColumns = retainedJoinColumnNames.Select(name => retainedDataFrame.Columns[name]).ToArray();
```

Comment:
Is `retainJoinColumns` here the same as just writing a for loop over `retainedJoinColumnNames`? If so, we should get rid of LINQ here to be performant.

Reply:
`retainJoinColumns` is a collection of columns, that are used for joining. It is used in `IsAnyNullValueInColumns` method for each row in the loop to get if there is any null value at least of one of the column, that is used as part of the key used for joining